File: src/kernellens/knowledge/schema.py
"""Stable, JSON-serializable semantic units and provenance."""

import hashlib
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

INDEXES = ("api", "concept", "example", "compiler", "operator")
SCHEMA_VERSION = 1


def sha256(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


class SourceLocation(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)
    path: str
    start_line: int = Field(ge=1)
    end_line: int = Field(ge=1)


class KnowledgeUnit(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)
    id: str
    category: Literal[
        "api", "instruction", "concept", "operator", "compiler", "example"
    ]
    name: str
    description: str
    when_to_use: str
    parameters: list[dict] = Field(default_factory=list)
    returns: str = ""
    examples: list[dict] = Field(default_factory=list)
    related_concepts: list[str] = Field(default_factory=list)
    source_location: SourceLocation
    keywords: list[str] = Field(default_factory=list)
    indexes: list[Literal["api", "concept", "example", "compiler", "operator"]]
    layer: Literal["user", "implementation"]
    visibility: Literal["public", "module", "internal", "example", "documentation"]
    aliases: list[str] = Field(default_factory=list)
    signature: str = ""
    symbols: list[str] = Field(default_factory=list)
    targets: list[str] = Field(default_factory=list)
    evidence_quality: Literal["source", "documentation", "example", "structural"]
    source_revision: str
    source_hash: str
    content: str
    context: str = ""
    cautions: list[str] = Field(default_factory=list)
    constraints: list[str] = Field(default_factory=list)
    hardware_mapping: list[str] = Field(default_factory=list)
    operator_structure: dict = Field(default_factory=dict)
    dependencies: list[str] = Field(default_factory=list)


def unit_id(category: str, path: str, name: str) -> str:
    # Lines are deliberately excluded: unrelated lines moving do not change identity.
    return "K" + sha256(f"{category}:{path}:{name}".encode())[:16]

File: src/knowledge/schema.ts
/**
 * Knowledge unit schema — stable, JSON-serializable semantic units and provenance.
 * Mirrors python-legacy/kernellens/knowledge/schema.py.
 */

import { createHash } from 'node:crypto';
import { z } from 'zod';

export const SCHEMA_VERSION = 1;
export const INDEXES = ['api', 'concept', 'example', 'compiler', 'operator'] as const;
export type IndexName = (typeof INDEXES)[number];

// ── SourceLocation ────────────────────────────────────────────────────────────

export const SourceLocation = z.object({
  path: z.string(),
  start_line: z.number().int().min(1),
  end_line: z.number().int().min(1),
});
export type SourceLocation = z.infer<typeof SourceLocation>;

// ── KnowledgeUnit ─────────────────────────────────────────────────────────────

export const Category = z.enum([
  'api',
  'instruction',
  'concept',
  'operator',
  'compiler',
  'example',
]);
export type Category = z.infer<typeof Category>;

export const KnowledgeUnit = z.object({
  id: z.string(),
  category: Category,
  name: z.string(),
  description: z.string(),
  when_to_use: z.string(),
  parameters: z.array(z.record(z.unknown())).default([]),
  returns: z.string().default(''),
  examples: z.array(z.record(z.unknown())).default([]),
  related_concepts: z.array(z.string()).default([]),
  source_location: SourceLocation,
  keywords: z.array(z.string()).default([]),
  indexes: z.array(z.enum(INDEXES)),
  layer: z.enum(['user', 'implementation']),
  visibility: z.enum(['public', 'module', 'internal', 'example', 'documentation']),
  aliases: z.array(z.string()).default([]),
  signature: z.string().default(''),
  symbols: z.array(z.string()).default([]),
  targets: z.array(z.string()).default([]),
  evidence_quality: z.enum(['source', 'documentation', 'example', 'structural']),
  source_revision: z.string(),
  source_hash: z.string(),
  content: z.string(),
  context: z.string().default(''),
  cautions: z.array(z.string()).default([]),
  constraints: z.array(z.string()).default([]),
  hardware_mapping: z.array(z.string()).default([]),
  operator_structure: z.record(z.unknown()).default({}),
  dependencies: z.array(z.string()).default([]),
});
export type KnowledgeUnit = z.infer<typeof KnowledgeUnit>;

// ── Helper ────────────────────────────────────────────────────────────────────

/**
 * Generate a deterministic unit ID from category, path, and name.
 * Mirrors Python: "K" + sha256(f"{category}:{path}:{name}")[:16]
 */
export function unitId(category: string, path: string, name: string): string {
  const h = createHash('sha256').update(`${category}:${path}:${name}`).digest('hex');
  return `K${h.slice(0, 16)}`;
}
